Bug: Compiling for arm64, error: cannot convert ‘const uint32x4_t’ to ‘uint8x16_t’ and similar errors

### What happened?

I'm compiling on free tier Oracle Ampere A1 server, arm64 architecture.
Running `make`, I'm getting a very long set of compiler errors, almost all are some form of can't convert uint to int, e.g.
- ggml/src/iqk/iqk_mul_mat.cpp:10303:36: error: cannot convert ‘const uint32x4_t’ to ‘uint8x16_t’
- ggml/src/iqk/iqk_mul_mat.cpp:10305:54: error: cannot convert ‘uint8x16_t’ to ‘int8x16_t’
- ggml/src/iqk/iqk_mul_mat.cpp:10334:40: error: cannot convert ‘int8x16_t’ to ‘__Uint8x16_t’ in assignment
- ggml/src/iqk/iqk_mul_mat.cpp:10954:41: error: cannot convert ‘const __Int16x8_t’ to ‘const int8x16_t&’
- ggml/src/iqk/iqk_mul_mat.cpp:11271:37: error: cannot convert ‘int8x16_t’ to ‘uint8x16_t’
... and many more in the full log below
[logs.txt](https://github.com/user-attachments/files/19582579/logs.txt)

I would expect it to compile with no errors.

### Name and Version

Git commit id
```
❯ git rev-parse HEAD
07dbc1aa06d761634419759431ebb215baf698bb
```

### What operating system are you seeing the problem on?

Linux
```
❯ uname -a
Linux instance-20240214-1712 6.8.0-1018-oracle #19~22.04.1-Ubuntu SMP Mon Dec  9 23:49:53 UTC 2024 aarch64 aarch64 aarch64 GNU/Linux
```

### Relevant log output

[logs.txt](https://github.com/user-attachments/files/19582579/logs.txt)

Still errors with int and float conversions, e.g. 
```
ggml/src/iqk/iqk_mul_mat.cpp:12791:44: error: cannot convert ‘int32x4_t’ to ‘float32x4_t’
```
[logs.txt](https://github.com/user-attachments/files/19582935/logs.txt)

I also tried adding `-fpermissive`, errors with the same

Not sure if this is an issue with just my setup
I'm getting 
```
/usr/bin/ld: ggml/src/ggml.o: in function `ggml_compute_forward_flash_attn_ext_f16':
ggml.c:(.text+0x83f0): undefined reference to `iqk_flash_attn_noalibi'
```
full log
```
❯ make CXXFLAGS="-fpermissive -flax-vector-conversions" CFLAGS="-flax-vector-conversions"
I ccache not found. Consider installing it for faster compilation.
make: jetson_release: No such file or directory
I llama.cpp build info: 
I UNAME_S:   Linux
I UNAME_P:   aarch64
I UNAME_M:   aarch64
I CFLAGS:    -Iggml/include -Iggml/src -Iinclude -Isrc -Icommon -D_XOPEN_SOURCE=600 -D_GNU_SOURCE -DNDEBUG -DGGML_USE_OPENMP -DGGML_USE_IQK_MULMAT -DGGML_USE_LLAMAFILE  -std=c11   -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wshadow -Wstrict-prototypes -Wpointer-arith -Wmissing-prototypes -Werror=implicit-int -Werror=implicit-function-declaration -pthread -mcpu=native -fopenmp -DGGML_USE_IQK_MULMAT -Wdouble-promotion -flax-vector-conversions
I CXXFLAGS:  -std=c++17 -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -pthread -mcpu=native -fopenmp -fpermissive -flax-vector-conversions  -Wno-array-bounds -Wno-format-truncation -Wextra-semi -Iggml/include -Iggml/src -Iinclude -Isrc -Icommon -D_XOPEN_SOURCE=600 -D_GNU_SOURCE -DNDEBUG -DGGML_USE_OPENMP -DGGML_USE_IQK_MULMAT -DGGML_USE_LLAMAFILE 
I NVCCFLAGS: -std=c++17 -O3 
I LDFLAGS:    
I CC:        cc (Ubuntu 12.3.0-1ubuntu1~22.04) 12.3.0
I CXX:       c++ (Ubuntu 12.3.0-1ubuntu1~22.04) 12.3.0

c++ -std=c++17 -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -pthread -mcpu=native -fopenmp -fpermissive -flax-vector-conversions  -Wno-array-bounds -Wno-format-truncation -Wextra-semi -Iggml/include -Iggml/src -Iinclude -Isrc -Icommon -D_XOPEN_SOURCE=600 -D_GNU_SOURCE -DNDEBUG -DGGML_USE_OPENMP -DGGML_USE_IQK_MULMAT -DGGML_USE_LLAMAFILE  -c examples/baby-llama/baby-llama.cpp -o examples/baby-llama/baby-llama.o
c++ -std=c++17 -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -pthread -mcpu=native -fopenmp -fpermissive -flax-vector-conversions  -Wno-array-bounds -Wno-format-truncation -Wextra-semi -Iggml/include -Iggml/src -Iinclude -Isrc -Icommon -D_XOPEN_SOURCE=600 -D_GNU_SOURCE -DNDEBUG -DGGML_USE_OPENMP -DGGML_USE_IQK_MULMAT -DGGML_USE_LLAMAFILE  ggml/src/iqk/iqk_quantize.o ggml/src/iqk/iqk_mul_mat.o ggml/src/llamafile/sgemm.o ggml/src/ggml.o ggml/src/ggml-alloc.o ggml/src/ggml-backend.o ggml/src/ggml-quants.o ggml/src/ggml-aarch64.o src/llama.o src/llama-vocab.o src/llama-grammar.o src/llama-sampling.o src/unicode.o src/unicode-data.o common/common.o common/console.o common/ngram-cache.o common/sampling.o common/train.o common/grammar-parser.o common/build-info.o common/json-schema-to-grammar.o examples/baby-llama/baby-llama.o -o llama-baby-llama  
/usr/bin/ld: ggml/src/ggml.o: in function `ggml_compute_forward_flash_attn_ext_f16':
ggml.c:(.text+0x83f0): undefined reference to `iqk_flash_attn_noalibi'
collect2: error: ld returned 1 exit status
make: *** [Makefile:1376: llama-baby-llama] Error 1
```

Hmm, getting other unresolved references with `cmake`
```
❯ cmake -B build -DCMAKE_CXX_FLAGS="-fpermissive -flax-vector-conversions" -DCMAKE_C_FLAGS="-flax-vector-conversions" && cmake --b
uild build --config Release 
-- OpenMP found
-- Using optimized iqk matrix multiplications
-- Using llamafile
-- Warning: ccache not found - consider installing it for faster compilation or disable this warning with GGML_CCACHE=OFF
-- CMAKE_SYSTEM_PROCESSOR: aarch64
-- ARM detected
-- Configuring done (0.2s)
-- Generating done (0.2s)
-- Build files have been written to: /home/ubuntu/projects/oobabooga_linux/ik_llama.cpp/build
[  6%] Built target ggml
[  7%] Building CXX object src/CMakeFiles/llama.dir/llama.cpp.o
[  7%] Building CXX object src/CMakeFiles/llama.dir/unicode-data.cpp.o
[  8%] Linking CXX shared library libllama.so
[ 10%] Built target llama
[ 11%] Built target build_info
[ 11%] Building CXX object common/CMakeFiles/common.dir/common.cpp.o
[ 12%] Building CXX object common/CMakeFiles/common.dir/sampling.cpp.o
[ 12%] Building CXX object common/CMakeFiles/common.dir/console.cpp.o
[ 13%] Building CXX object common/CMakeFiles/common.dir/grammar-parser.cpp.o
[ 14%] Building CXX object common/CMakeFiles/common.dir/json-schema-to-grammar.cpp.o
[ 14%] Building CXX object common/CMakeFiles/common.dir/train.cpp.o
[ 15%] Building CXX object common/CMakeFiles/common.dir/ngram-cache.cpp.o
[ 15%] Linking CXX static library libcommon.a
[ 15%] Built target common
[ 16%] Building CXX object tests/CMakeFiles/test-tokenizer-0.dir/test-tokenizer-0.cpp.o
[ 17%] Linking CXX executable ../bin/test-tokenizer-0
/usr/bin/ld: ../ggml/src/libggml.so: undefined reference to `iqk_moe_fused_up_gate'
/usr/bin/ld: ../ggml/src/libggml.so: undefined reference to `iqk_mul_mat_4d'
/usr/bin/ld: ../ggml/src/libggml.so: undefined reference to `iqk_mul_mat'
/usr/bin/ld: ../ggml/src/libggml.so: undefined reference to `iqk_flash_attn_noalibi'
/usr/bin/ld: ../ggml/src/libggml.so: undefined reference to `iqk_mul_mat_moe'
collect2: error: ld returned 1 exit status
gmake[2]: *** [tests/CMakeFiles/test-tokenizer-0.dir/build.make:103: bin/test-tokenizer-0] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:2123: tests/CMakeFiles/test-tokenizer-0.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
```

I'm using whatever the default settings are.
Adding `-DGGML_NATIVE=1`, running the following unfortunately still errors
```
❯ cmake -B build -DGGML_NATIVE=1 -DCMAKE_CXX_FLAGS="-fpermissive -flax-vector-conversions" -DCMAKE_C_FLAGS="-flax-vector-conversions" && cmake --build build --config Release -j 4
-- The C compiler identification is GNU 12.3.0
-- The CXX compiler identification is GNU 12.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: /usr/bin/git (found version "2.34.1")
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Found OpenMP_C: -fopenmp (found version "4.5")
-- Found OpenMP_CXX: -fopenmp (found version "4.5")
-- Found OpenMP: TRUE (found version "4.5")
-- OpenMP found
-- Using optimized iqk matrix multiplications
-- Using llamafile
-- Warning: ccache not found - consider installing it for faster compilation or disable this warning with GGML_CCACHE=OFF
-- CMAKE_SYSTEM_PROCESSOR: aarch64
-- ARM detected
-- Performing Test COMPILER_SUPPORTS_FP16_FORMAT_I3E
-- Performing Test COMPILER_SUPPORTS_FP16_FORMAT_I3E - Failed
-- Configuring done (1.9s)
-- Generating done (0.2s)
-- Build files have been written to: /home/ubuntu/projects/oobabooga_linux/ik_llama.cpp/build
[  1%] Building C object examples/gguf-hash/CMakeFiles/sha256.dir/deps/sha256/sha256.c.o
[  2%] Building C object examples/gguf-hash/CMakeFiles/xxhash.dir/deps/xxhash/xxhash.c.o
[  3%] Building C object ggml/src/CMakeFiles/ggml.dir/ggml.c.o
[  4%] Building CXX object common/CMakeFiles/build_info.dir/build-info.cpp.o
[  4%] Built target build_info
[  4%] Building C object ggml/src/CMakeFiles/ggml.dir/ggml-alloc.c.o
[  4%] Built target sha256
[  5%] Building C object ggml/src/CMakeFiles/ggml.dir/ggml-backend.c.o
[  6%] Building C object ggml/src/CMakeFiles/ggml.dir/ggml-quants.c.o
[  6%] Building CXX object ggml/src/CMakeFiles/ggml.dir/llamafile/sgemm.cpp.o
[  6%] Built target xxhash
[  6%] Building C object examples/gguf-hash/CMakeFiles/sha1.dir/deps/sha1/sha1.c.o
In function ‘SHA1Update’,
    inlined from ‘SHA1Final’ at /home/ubuntu/projects/oobabooga_linux/ik_llama.cpp/examples/gguf-hash/deps/sha1/sha1.c:265:5:
/home/ubuntu/projects/oobabooga_linux/ik_llama.cpp/examples/gguf-hash/deps/sha1/sha1.c:219:13: warning: ‘SHA1Transform’ reading 64 bytes from a region of size 0 [-Wstringop-overread]
  219 |             SHA1Transform(context->state, &data[i]);
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ubuntu/projects/oobabooga_linux/ik_llama.cpp/examples/gguf-hash/deps/sha1/sha1.c:219:13: note: referencing argument 2 of type ‘const unsigned char[64]’
/home/ubuntu/projects/oobabooga_linux/ik_llama.cpp/examples/gguf-hash/deps/sha1/sha1.c: In function ‘SHA1Final’:
/home/ubuntu/projects/oobabooga_linux/ik_llama.cpp/examples/gguf-hash/deps/sha1/sha1.c:54:6: note: in a call to function ‘SHA1Transform’
   54 | void SHA1Transform(
      |      ^~~~~~~~~~~~~
In function ‘SHA1Update’,
    inlined from ‘SHA1Final’ at /home/ubuntu/projects/oobabooga_linux/ik_llama.cpp/examples/gguf-hash/deps/sha1/sha1.c:269:9:
/home/ubuntu/projects/oobabooga_linux/ik_llama.cpp/examples/gguf-hash/deps/sha1/sha1.c:219:13: warning: ‘SHA1Transform’ reading 64 bytes from a region of size 0 [-Wstringop-overread]
  219 |             SHA1Transform(context->state, &data[i]);
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ubuntu/projects/oobabooga_linux/ik_llama.cpp/examples/gguf-hash/deps/sha1/sha1.c:219:13: note: referencing argument 2 of type ‘const unsigned char[64]’
/home/ubuntu/projects/oobabooga_linux/ik_llama.cpp/examples/gguf-hash/deps/sha1/sha1.c: In function ‘SHA1Final’:
/home/ubuntu/projects/oobabooga_linux/ik_llama.cpp/examples/gguf-hash/deps/sha1/sha1.c:54:6: note: in a call to function ‘SHA1Transform’
   54 | void SHA1Transform(
      |      ^~~~~~~~~~~~~
[  6%] Built target sha1
[  7%] Building CXX object ggml/src/CMakeFiles/ggml.dir/iqk/iqk_mul_mat.cpp.o
/home/ubuntu/projects/oobabooga_linux/ik_llama.cpp/ggml/src/iqk/iqk_mul_mat.cpp:17501:6: warning: no previous declaration for ‘bool iqk_mul_mat(int, long int, long int, long int, int, const void*, long int, int, const void*, long int, float*, long int, int, int)’ [-Wmissing-declarations]
17501 | bool iqk_mul_mat(int, long, long, long, int, const void *, long, int, const void *, long, float *, long, int, int) {
      |      ^~~~~~~~~~~
/home/ubuntu/projects/oobabooga_linux/ik_llama.cpp/ggml/src/iqk/iqk_mul_mat.cpp:17505:6: warning: no previous declaration for ‘bool iqk_mul_mat_4d(long int, long int, long int, long int, long int, long int, long int, long int, long int, long int, long int, long int, long int, int, const void*, long int, int, const void*, long int, float*, long int, int, int)’ [-Wmissing-declarations]
17505 | bool iqk_mul_mat_4d(long /*Nx*/, long /*Ny*/, long /*ne00*/,
      |      ^~~~~~~~~~~~~~
/home/ubuntu/projects/oobabooga_linux/ik_llama.cpp/ggml/src/iqk/iqk_mul_mat.cpp:17514:6: warning: no previous declaration for ‘bool iqk_mul_mat_moe(long int, long int, long int, int, int, const void*, long int, int, const void*, long int, float*, long int, long int, const void*, int, int)’ [-Wmissing-declarations]
17514 | bool iqk_mul_mat_moe(long, long, long, int, int, const void *, long, int, const void *, long, float *, long, long,
      |      ^~~~~~~~~~~~~~~
/home/ubuntu/projects/oobabooga_linux/ik_llama.cpp/ggml/src/iqk/iqk_mul_mat.cpp:17519:6: warning: no previous declaration for ‘bool iqk_moe_fused_up_gate(long int, long int, long int, int, int, int, const void*, const void*, long int, int, const void*, long int, float*, long int, long int, const void*, int, int)’ [-Wmissing-declarations]
17519 | bool iqk_moe_fused_up_gate(long /*Nx*/, long /*Ny*/, long /*ne00*/, int /*ne11*/, int /*unary_op*/,
      |      ^~~~~~~~~~~~~~~~~~~~~
[  7%] Building CXX object ggml/src/CMakeFiles/ggml.dir/iqk/iqk_flash_attn.cpp.o
/home/ubuntu/projects/oobabooga_linux/ik_llama.cpp/ggml/src/iqk/iqk_flash_attn.cpp:189:6: warning: no previous declaration for ‘bool iqk_flash_attn_noalibi(int, int, float, int, int, long int, long int, int, int, long int, long int, int, int, long int, long int, int, int, long int, int, int, int, int, int, int, int, int, int, const void*, const void*, const void*, const void*, float, float, float*, void*, barrier_t, void*, int, int)’ [-Wmissing-declarations]
  189 | bool iqk_flash_attn_noalibi([[maybe_unused]] int type_q, [[maybe_unused]] int type_mask, [[maybe_unused]] float max_bias,
      |      ^~~~~~~~~~~~~~~~~~~~~~
[  8%] Building CXX object ggml/src/CMakeFiles/ggml.dir/iqk/iqk_quantize.cpp.o
[  8%] Building C object ggml/src/CMakeFiles/ggml.dir/ggml-aarch64.c.o
[  9%] Linking CXX shared library libggml.so
[  9%] Built target ggml
[ 10%] Building CXX object examples/gguf-hash/CMakeFiles/llama-gguf-hash.dir/gguf-hash.cpp.o
[ 10%] Building CXX object src/CMakeFiles/llama.dir/llama-vocab.cpp.o
[ 11%] Building CXX object src/CMakeFiles/llama.dir/llama.cpp.o
[ 11%] Building CXX object examples/gguf/CMakeFiles/llama-gguf.dir/gguf.cpp.o
[ 12%] Linking CXX executable ../../bin/llama-gguf
[ 12%] Linking CXX executable ../../bin/llama-gguf-hash
/usr/bin/ld: ../../ggml/src/libggml.so: undefined reference to `iqk_moe_fused_up_gate'
/usr/bin/ld: ../../ggml/src/libggml.so: undefined reference to `iqk_mul_mat_4d'
/usr/bin/ld: ../../ggml/src/libggml.so: undefined reference to `iqk_mul_mat'
/usr/bin/ld: ../../ggml/src/libggml.so: undefined reference to `iqk_flash_attn_noalibi'
/usr/bin/ld: ../../ggml/src/libggml.so: undefined reference to `iqk_mul_mat_moe'
collect2: error: ld returned 1 exit status
gmake[2]: *** [examples/gguf/CMakeFiles/llama-gguf.dir/build.make:101: bin/llama-gguf] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:3196: examples/gguf/CMakeFiles/llama-gguf.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
[ 13%] Building CXX object src/CMakeFiles/llama.dir/llama-grammar.cpp.o
/usr/bin/ld: ../../ggml/src/libggml.so: undefined reference to `iqk_moe_fused_up_gate'
/usr/bin/ld: ../../ggml/src/libggml.so: undefined reference to `iqk_mul_mat_4d'
/usr/bin/ld: ../../ggml/src/libggml.so: undefined reference to `iqk_mul_mat'
/usr/bin/ld: ../../ggml/src/libggml.so: undefined reference to `iqk_flash_attn_noalibi'
/usr/bin/ld: ../../ggml/src/libggml.so: undefined reference to `iqk_mul_mat_moe'
collect2: error: ld returned 1 exit status
gmake[2]: *** [examples/gguf-hash/CMakeFiles/llama-gguf-hash.dir/build.make:107: bin/llama-gguf-hash] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:3038: examples/gguf-hash/CMakeFiles/llama-gguf-hash.dir/all] Error 2
[ 13%] Building CXX object src/CMakeFiles/llama.dir/llama-sampling.cpp.o
[ 14%] Building CXX object src/CMakeFiles/llama.dir/unicode.cpp.o
[ 14%] Building CXX object src/CMakeFiles/llama.dir/unicode-data.cpp.o
[ 15%] Linking CXX shared library libllama.so
[ 15%] Built target llama
gmake: *** [Makefile:146: all] Error 2
```

--- a/ggml/src/iqk/iqk_mul_mat.cpp
+++ b/ggml/src/iqk/iqk_mul_mat.cpp
@@ -15829,9 +15829,9 @@ struct FlashMS {
             auto vm2 = vzip2q_u16(vm, vm);
             auto kq  = vld1q_f32_x2(cache + k_step*j + 8*l);
             vk[2*l+0] = vreinterpretq_f32_u32(vorrq_u32(vandq_u32(vreinterpretq_u32_f32(kq.val[0]), vm1),
-                                                        vbicq_u32(vinf, vm1)));
+                                                        vbicq_u32(vreinterpretq_u32_f32(vinf), vm1)));
             vk[2*l+1] = vreinterpretq_f32_u32(vorrq_u32(vandq_u32(vreinterpretq_u32_f32(kq.val[1]), vm2),
-                                                        vbicq_u32(vinf, vm2)));
+                                                        vbicq_u32(vreinterpretq_u32_f32(vinf), vm2)));
         }
         float32x4_t vmax = vdupq_n_f32(-INFINITY);
         auto vscale32 = vcvt_f32_f16(vget_low_f16(vscale));

--- a/ggml/src/iqk/iqk_mul_mat.cpp
+++ b/ggml/src/iqk/iqk_mul_mat.cpp
@@ -12735,7 +12735,7 @@ static void mul_mat_iq1_s_r4_q8_1(int n, const void * vx, size_t bx, const DataI
     int nb = n / 32;
     GGML_ASSERT(nb%4 == 0);
     uint8x16_t qx[8];
-    int32x4_t acc[nrc_y] = {};
+    float32x4_t acc[nrc_y] = {};
     auto ms = vdup_n_u16(0x8000);
     auto mask = vdupq_n_s8(0x03);
     float d8[4*nrc_y];
@@ -14140,7 +14140,7 @@ void mul_mat_q8_KV_r8_q8_KV(int n, const void * vx, size_t bx, const DataInfo& i
             auto scale2 = vmulq_f32(scale2_x, scale_y);
             info.store(ix+0, iy, vmulq_f32(scale1, vcvtq_f32_s32(acc[2*iy+0])));
             info.store(ix+4, iy, vmulq_f32(scale2, vcvtq_f32_s32(acc[2*iy+1])));
-            acc[2*iy+0] = acc[2*iy+1] = vdupq_n_f32(0.f);
+            acc[2*iy+0] = acc[2*iy+1] = vdupq_n_s32(0.f);
         }
     }
 }
@@ -14823,11 +14823,11 @@ inline float32x4_t v_tanh(float32x4_t x) {
     return vreinterpretq_f32_u32(vorrq_u32(vandq_u32(vreinterpretq_u32_f32(one), mask), vbicq_u32(vreinterpretq_u32_f32(res), mask)));
     //return vdivq_f32(vsubq_f32(exp_two_x, one), vaddq_f32(exp_two_x, one));
 }
-inline float32x4_t v_tanh(float16x8_t x) {
-    auto val1 = v_tanh(vcvt_f32_f16(vget_low_f16(x)));
-    auto val2 = v_tanh(vcvt_f32_f16(vget_high_f16(x)));
-    return vcombine_f16(vcvt_f16_f32(val1), vcvt_f16_f32(val2));
-}
+//inline float32x4_t v_tanh(float16x8_t x) {
+//    auto val1 = v_tanh(vcvt_f32_f16(vget_low_f16(x)));
+//    auto val2 = v_tanh(vcvt_f32_f16(vget_high_f16(x)));
+//    return vcombine_f16(vcvt_f16_f32(val1), vcvt_f16_f32(val2));
+//}
 inline float32x4_t v_silu(float32x4_t x) {
     const float32x4_t one = vdupq_n_f32(1.0f);
     const float32x4_t zero = vdupq_n_f32(0.0f);
@@ -15671,7 +15671,9 @@ struct HelperQ60 final : public BaseHelper<step> {
         auto dl = (const block_q6_0 *)Base::lblock(l1) + j/QK6_0;
 #ifdef __aarch64__
         // TODO
-        auto vd = F16::set1(*(const float16_t *)&dl->d);
+        const float16_t * d16 = (const float16_t *)&dl->d;
+        auto vd = F16::set1(d16[0]);
+        //auto vd = F16::set1(*(const float16_t *)&dl->d);
         auto qh8 = vld1_u8(dl->qh);
         auto qh  = vcombine_u8(vshl_n_u8(qh8, 4), qh8);
         auto qs  = vld1q_u8(dl->qs);
@@ -15819,7 +15821,7 @@ struct FlashMS {
         return vmaxvq_f32(vmax);
     }
     inline float load_apply_mask_and_scale(int j, float32x4_t * vk, const char * mask) {
-        auto vzero = vdupq_n_f32(0);
+        auto vzero = vdupq_n_f16(0);
         auto vinf  = vdupq_n_f32(-INFINITY);
         for (int l = 0; l < k_step/8; ++l) {
             auto vm = vceqq_f16(vzero, vld1q_f16((const float16_t *)mask + 8*l));

